### PR TITLE
[Do not Merge] fix some pipeline issues - WIP

### LINKED ISSUE #2064: ls | more does not work on mac
xonsh version: 0.5.2
python version 3.6.0

*****************************************************************************************************************************
```
>>> ls | more
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 17380, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 1964, in run_compiled_code
    func(code, glb, loc)
  File "<xonsh-code>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 16159, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 16127, in run_subproc
    command.end()
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 11116, in end
    for _ in self.tee_stdout():
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 11041, in tee_stdout
    for line in self.iterraw():
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 10940, in iterraw
    wait_for_active_job()
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 3694, in wait_for_active_job
    _continue(active_task)
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 3607, in _continue
    _send_signal(job, signal.SIGCONT)
  File "/usr/local/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 3542, in _send_signal
    os.kill(pid, signal)
ProcessLookupError: [Errno 3] No such process
```
*****************************************************************************************************************************

Also, I created a alias function ls-more on python version 3.5... now that does not work either.

def _lsmore(args, stdin):                                                                                                           
    ls -lt @(args) >> .ls_more                                                                                                      
    more .ls_more                                                                                                                   
    rm .ls_more   

aliases['ls-more'] = _lsmore

### PR DESCRIPTION
**Pull request by mitnk**
This PR try fixes #2064, and continue the work of #2065. Here are [some investigation results](https://github.com/xonsh/xonsh/pull/2065#issuecomment-276565675).

This PR is still WIP. I need more tests on Unix/Linux platforms, and more changes may followed.

### CODE REVIEW

**Comment by scopatz** [comment]
I don't know why travis is claiming this fails when it doesn't. Thanks for putting this in @mitnk!

File: xonsh/built_ins.py [hunk 1]
```diff
@@ -540,7 +539,6 @@ def prep_preexec_fn(self, kwargs, pipeline_group=None):
         else:
             def xonsh_preexec_fn():
                 """Preexec function bound to a pipeline group."""
-                os.setpgid(0, pipeline_group)
```
Comment (scopatz) [inline xonsh/built_ins.py]: I think we want to keep this around for Linux.  Also maybe add a comment that we aren't sure why it doesn't work on Mac.

File: xonsh/built_ins.py [hunk 2]
```diff
@@ -345,7 +345,6 @@ def no_pg_xonsh_preexec_fn():
     """Default subprocess preexec function for when there is no existing
     pipeline group.
     """
-    os.setpgrp()
```
Comment (scopatz) [inline xonsh/built_ins.py]: Same as below, I think we want to keep this on linux, and perhaps add a comment that we don't know why it doesn't work on mac.

**Comment by mitnk** [comment]
hi @scopatz Yes, only remove them on Mac was my idea too. But then I think maybe we don't need to create a process group to make job control works, then why not? :)

While I was checking the original issue, I found this issue should be a generic issue, not a Mac only. For example, if the first process run & finish very fast, then the `os.setpgid(0, pipeline_group)` will raise errors, because the it was the first process and it's already gone. The following is a evil diff for this:
```
diff --git a/xonsh/built_ins.py b/xonsh/built_ins.py
index 032e197..0b287d9 100644
--- a/xonsh/built_ins.py
+++ b/xonsh/built_ins.py
@@ -795,6 +795,7 @@ def run_subproc(cmds, captured=False):
         starttime = time.time()
         proc = spec.run(pipeline_group=pipeline_group)
         procs.append(proc)
+        time.sleep(2.0)
         if ON_POSIX and pipeline_group is None and \
            spec.cls is subprocess.Popen:
             pipeline_group = proc.pid
diff --git a/xonsh/jobs.py b/xonsh/jobs.py
index d2e296b..57c9f1d 100644
--- a/xonsh/jobs.py
+++ b/xonsh/jobs.py
@@ -115,7 +115,7 @@ else:
     _shell_pgrp = os.getpgrp()

     _block_when_giving = LazyObject(lambda: (signal.SIGTTOU, signal.SIGTTIN,
-                                             signal.SIGTSTP, signal.SIGCHLD),
+                                             signal.SIGTSTP),
                                     globals(), '_block_when_giving')

     # check for shell tty
diff --git a/xonsh/main.py b/xonsh/main.py
index ac06772..b1f0976 100644
--- a/xonsh/main.py
+++ b/xonsh/main.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 """The main xonsh script."""
 import os
+import signal
 import sys
 import enum
 import argparse
@@ -298,6 +299,9 @@ def main(argv=None):

 def main_xonsh(args):
     """Main entry point for xonsh cli."""
+    def func_child(*args, **kwargs):
+        os.waitpid(-1, os.WNOHANG)
+    signal.signal(signal.SIGCHLD, func_child)
     events.on_post_init.fire()
     env = builtins.__xonsh_env__
     shell = builtins.__xonsh_shell__
```
With this diff, even on Linux, the `ls | more` will raise exception:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/xonsh/base_shell.py", line 322, in default
    run_compiled_code(code, self.ctx, None, 'single')
  File "/usr/local/lib/python3.6/site-packages/xonsh/codecache.py", line 65, in run_compiled_code
    func(code, glb, loc)
  File "/home/pi/.xonsh/xonshrc.xsh", line 1, in <module>
    $XONSH_DEBUG = 1
  File "/usr/local/lib/python3.6/site-packages/xonsh/built_ins.py", line 860, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/local/lib/python3.6/site-packages/xonsh/built_ins.py", line 796, in run_subproc
    proc = spec.run(pipeline_group=pipeline_group)
  File "/usr/local/lib/python3.6/site-packages/xonsh/built_ins.py", line 501, in run
    p = self._run_binary(kwargs)
  File "/usr/local/lib/python3.6/site-packages/xonsh/built_ins.py", line 511, in _run_binary
    p = self.cls(self.cmd, bufsize=bufsize, **kwargs)
  File "/usr/local/lib/python3.6/subprocess.py", line 707, in __init__
    restore_signals, start_new_session)
  File "/usr/local/lib/python3.6/subprocess.py", line 1327, in _execute_child
    raise child_exception_type(err_msg)
subprocess.SubprocessError: Exception occurred in preexec_fn.
```
We could `try-except` the `os.setpgid(0, pipeline_group)` line though. But then it will fail over to a situation where "process group" is not used.

Anyway, this PR could eventually only make Mac do not create process group, but I'd like spend some time to try some thing in general first :)

**Comment by mitnk** [comment]
I may found a real case related create process group **on Linux** (Unix too) - 

On latest master do:

```
$ history info | more
^Z  # <-- try a few times more if not raise errors

$ history info | more
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/xonsh/main.py", line 294, in main
    return main_xonsh(args)
  File "/usr/local/lib/python3.6/site-packages/xonsh/main.py", line 317, in main_xonsh
    shell.shell.cmdloop()
  File "/usr/local/lib/python3.6/site-packages/xonsh/ptk/shell.py", line 144, in cmdloop
    line = self.singleline(auto_suggest=auto_suggest)
  File "/usr/local/lib/python3.6/site-packages/xonsh/ptk/shell.py", line 111, in singleline
    line = self.prompter.prompt(**prompt_args)
  File "/usr/local/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 118, in prompt
    eventloop.close()
  File "/usr/local/lib/python3.6/site-packages/prompt_toolkit/eventloop/posix.py", line 270, in close
    os.close(schedule_pipe[0])
OSError: [Errno 9] Bad file descriptor
Xonsh encountered an issue during launch
Failback to /bin/sh
```

On branch of this PR, `history info | more` works fine.

xonfig:
```
+------------------+-----------------------------------------------+
| xonsh            | 0.5.3.dev3263                                 |
| Git SHA          | 1a3d7ce                                       |
| Commit Date      | Feb 6 04:53:19 2017                           |
| Python           | 3.6.0                                         |
| PLY              | 3.9                                           |
| have readline    | True                                          |
| prompt toolkit   | 1.0.9                                         |
| shell type       | prompt_toolkit                                |
| pygments         | 2.1.3                                         |
| on posix         | True                                          |
| on linux         | <xonsh.lazyasd.LazyBool object at 0x765c6bf0> |
| distro           | debian                                        |
| on darwin        | False                                         |
| on windows       | False                                         |
| on cygwin        | False                                         |
| is superuser     | False                                         |
| default encoding | utf-8                                         |
| xonsh encoding   | utf-8                                         |
| encoding errors  | surrogateescape                               |
+------------------+-----------------------------------------------+
```

**Comment by scopatz** [comment]
That might not be the best example since `history` is a callable alias. The ProcProxy class has its own problems

**Comment by mitnk** [comment]
Right, the examples in my previous two comments are bot NOT good ones. I mean the direction of not creating process group (i.e. the current code in this PR) is not correct (even for Mac only).

Without creating a new process group, running simple command like `psql your-db` will in trouble - When hitting `Ctrl-C`, both psql and xonsh proc will receive SIGINT - which causes chaos.

So the correct way to fix the issue is to make the stopped (T) process continue in `wait_for_active_job()`.
```
 PID  PPID  PGID TPGID STAT COMMAND
1856  1821  1856  1856 S+   xonsh
1865  1856  1864  1856 T    less
```
[again, details here](https://github.com/xonsh/xonsh/pull/2065#issuecomment-276565675).

**Comment by mitnk** [comment]
@scopatz Updated a bit. I think it's fixed the issue [1] on Mac, though I feel it's not an elegant way.

It currently only fixes for Mac, but the "not-creating-process-group" and "handle SIGINT in xonsh" way also works on Linux too. The good part is it does need to manipulate `progress group and controlling terminals` things anymore. Not sure if there are any other effects.

[1] the only issue I found with this fix is a inside a `psql mydb` command, one `Ctrl-C` seems make multiple `SIGINT` to psql proc. I don't know the reason yet.

**Comment by mitnk** [comment]
Don't think this is the right thing to do. hold on for now

**Comment by scopatz** [comment]
Good think you said that when you did. I'll hold off for now.

**Comment by mitnk** [comment]
replaced by #2177

**Pull request closed without merging**